### ISSUE TITLE
example 2 won't finish 

### ISSUE THREAD

**Issue by evgenizer** [automl/HpBandSter#48]
Hi, 
thanks for the code, looks very interesting. 

I am trying to run a variant of example 2. I also use logging with 
result_logger = hpres.json_result_logger(directory=working_directory, overwrite=True)
that I pass to BOHB(...result_logger=result_logger,...) 

There are 2 problems:
1. The logger produces config.json and results.json but the latter will typically miss results for all but one initial [0,num_workers) configurations, but will show results for [num_workers, n_iterations) configurations.
2. After the optimizer goes over all configurations it just hangs, waiting for workers:

DEBUG:hpbandster:HBMASTER: Trying to run another job!
DEBUG:hpbandster:job_callback for (19, 0, 0) finished
DEBUG:hpbandster:DISPATCHER: Starting worker discovery
DEBUG:hpbandster:DISPATCHER: Found 8 potential workers, 8 currently in the pool.
DEBUG:hpbandster:DISPATCHER: Finished worker discovery
DEBUG:hpbandster:DISPATCHER: Starting worker discovery
DEBUG:hpbandster:DISPATCHER: Found 8 potential workers, 8 currently in the pool.
DEBUG:hpbandster:DISPATCHER: Finished worker discovery
DEBUG:hpbandster:DISPATCHER: Starting worker discovery
DEBUG:hpbandster:DISPATCHER: Found 8 potential workers, 8 currently in the pool.
DEBUG:hpbandster:DISPATCHER: Finished worker discovery
DEBUG:hpbandster:DISPATCHER: Starting worker discovery
DEBUG:hpbandster:DISPATCHER: Found 8 potential workers, 8 currently in the pool.
DEBUG:hpbandster:DISPATCHER: Finished worker discovery
DEBUG:hpbandster:DISPATCHER: Starting worker discovery
DEBUG:hpbandster:DISPATCHER: Found 8 potential workers, 8 currently in the pool.
DEBUG:hpbandster:DISPATCHER: Finished worker discovery
DEBUG:hpbandster:DISPATCHER: Starting worker discovery
............

and so on until I kill the processes.

Am I missing something: is there a way to exit gracefully?

Thanks!

**Comment by sfalkner** [comment]
Could you provide a bit more detail, please?
What's the OS, python version etc.

**Comment by evgenizer** [comment]
Ubuntu 16.04, python 3.7, part of anaconda 5.2.

Thanks

**Comment by sfalkner** [comment]
Hmmm...the original example runs through fine here. Have you changed anything else, e.g. the worker?

**Comment by evgenizer** [comment]
Yes, the worker is modified, it is problem-specific. The code runs well on my mac inside pycharm, but hangs in background runs on ubuntu.
Here is the fragment of “main” that calls the optimizer:

host = '127.0.0.1'
port = 9090


result_logger = hpres.json_result_logger(directory=working_directory, overwrite=True)


# Step 1: Start a nameserver
# Every run needs a nameserver. It could be a 'static' server with a
# permanent address, but here it will be started for the local machine with the default port.
# The nameserver manages the concurrent running workers across all possible threads or clusternodes.
# Note the run_id argument. This uniquely identifies a run of any HpBandSter optimizer.
NS = hpns.NameServer(run_id=run_id, host=host, port=None)
NS.start()

# Step 2: Start a worker
# Now we can instantiate a worker, providing the mandatory information
# Besides the sleep_interval, we need to define the nameserver information and
# the same run_id as above.
workers = []
for i in range(args.num_workers):
    worker = bohb_worker.BOHBWorker(args, device, nameserver=host, run_id=run_id, save_log=False, id=i)
    worker.run(background=True)
    workers.append(worker)




# Step 3: Run an optimizer
# Now we can create an optimizer object and start the run.
# Here, we run BOHB, but that is not essential.
# The run method will return the `Result` that contains all runs performed.
bohb = BOHB(configspace=worker.get_configspace(),
            run_id=run_id, nameserver=host, host=host,
            result_logger=result_logger,
            min_budget=args.min_budget, max_budget=args.max_budget,
            working_directory=working_directory
            )

result = bohb.run(n_iterations=args.num_configurations, min_n_workers=args.num_workers)


# Step 4: Shutdown
# After the optimizer run, we must shutdown the master and the nameserver.
bohb.shutdown(shutdown_workers=True)
NS.shutdown()

**Comment by sfalkner** [comment]
That code looks fine. You can try to run that with the example worker to make sure.
What does your worker return?

**Comment by evgenizer** [comment]
The worker returns dict={‘loss’:.., ‘info’:…}.

I was further looking at the analysis of the run, and at least in the multi-threaded setup the results are somewhat confusing. I got a few questions:
•  Job order vs. budget

  1.  Does BOHB start all runs at a given budget before starting jobs at the next higher budget?
  2.  Or does it simultaneously start jobs at multiple budgets?
•  Jobs per budget

  1.  How many jobs are run at each budget?
  2.  Is it C/budget for some constant C? (The data for the existing run does not follow this rule, e.g. third plot in  https://automl.github.io/HpBandSter/build/html/auto_examples/plot_example_6_analysis.html ).
•  Hyperparameter configuration reuse

  1.  Does BOHB sometimes reuse hyperparameter configurations for one budget in a run with a higher budget?
  2.  If yes, how does it decide when to do this?
  3.  Is the run with higher budget started from scratch, or does it start from the result of the training at the lower budget (i.e. a warm start?
•  Rank correlation

  1.  Is rank correlation based only on runs that used the exact same hyperparameters at two different budgets?
Thanks!
Evgeny

**Comment by sfalkner** [comment]
Hey Evgeny,
About the return value of your workers `compute` method. What are the entries in the info dict. If you try to return any non-build in types (like numpy arrays) the workers will hang and the run does not finish.

I will try to answer your questions, but for further detail, please refer to the paper.

• Job order vs. budget:
For each iteration, BOHB will start a certain number of configurations. For most iterations, that starting budget is smaller than the `max_budget`. After all of them finish, a third of them will then be evaluated on a higher budget (usually three times the previous budget). This is repeated until `max_budget` is reached. The strategy is called SuccessiveHalving (although SuccessiveThirding would be a better name ;) ) Depending on how many workers are available, BOHB will start running jobs on different budgets. In fact, most of the time most workers are busy evaluating a configuration on the biggest budget.

• Jobs per budget:
In each iteration there are roughly  3 times more evaluations from one budget to the next larger one. As different iterations start with different budgets, this factor does not hold exactly when you look at all iterations. I didn't do the math exactly, but one could determine the constant for each budget, but I is not essential. The plot you mention in you post does follow this trend of a constant number of evaluations per budget per time if you average out the effect of the iterations/limited number of workers. Those cause the 'staircase' look, but if you smooth that out, you should recover a linear trend for all budgets.

• Hyperparameter configuration reuse
Yes, configurations that performed well on the smaller budget will be evaluated on a larger budget (see above). By default, BOHB does not know about your computation, hence cannot save your intermediate training results and warmstart from there. Checkout this [issue](https://github.com/automl/HpBandSter/issues/42) for some more detail.

• Rank correlation
Yes, it's the only way to empirically measure how good of an indicator  the performance on a smaller budget is for the performance on larger budget.

Hope that helps.

**Comment by evgenizer** [comment]
OK, got it, some of the entries in returned dict were elements of numpy. Casting then to floats helps to solve problem 1 in the original post. Now I am testing if problem 2 is resolved also. 

I did read the paper before trying out the software :). Still confused though:

1. What do you call 'iteration': is it complete training with a number of configurations at a given budget? Or is it a complete training of a number of configurations across all budgets?
2. I understand the successive halving (SH) and the continuation of good 1/3 of configurations to a higher budget, but I thought that in the paper you discuss generating a set of configurations based on current model before moving to the next budget... Your description now seems to imply that fractions of the original configurations chosen at the lowest budget are propagated across budgets as in SH. Do you use the configurations that survived that race to the highest budget in order to update the model and pick up new set of configurations for next iteration at the lowest budget?     
3. The plot follows the trend, but the ratio between the highest to the lowest budget runs is ~1/2 not 1/9.

**Comment by sfalkner** [comment]
1. One iteration is one round of SuccessiveHalving, meaning a bunch of configuration on a small budget and then less and less configurations thereof evaluated on larger and larger budgets.
2. BOHB does use the original SH, but the sampling of random configurations is replaced by samples from a model. Every evaluation is used to update the model though.
3. Okay, let's do the math for a simple example. Let's say `min_budget=1' and 'max_budget=9'. Then the budgets are [1,3,9]. The first, fourth, seventh and so on iterations will run 9 configs on budget 1, 3 on 3 and 1 on 9. The second, fifth, eighth and so iteration will use 3 as the lowest budget and start with  3 or 5 configs (not sure right now) on budget 3 and have 1 on the largest budget. The third, sixth, ninth and so on iteration will have only 3 configuration on the maximum budget. If I sum that up:
9 configurations on the smallest budget and 1+1+3 = 5 configurations on the largest budgets, which is a ratio of 5/9 ~ 0.5.
Does that clarify things a bit?

**Comment by evgenizer** [comment]
OK, great. This does clarify, thanks. 

Number of iterations is the parameter passed to BOHB.run(n_iterations)?

3. In your example: I think the second iteration should have 5 configs on budget=3, based on Algorithm 1, line 3, in the paper, correct?  Line 4 in the same algorithms should have eta^{-s} b_max, shouldn't it?

**Comment by sfalkner** [comment]
I think the implementation [here](https://github.com/automl/HpBandSter/blob/master/hpbandster/optimizers/bohb.py#L119) will actually produce 3 configs in the second iteration. This is due to some confusion I had about the original HyperBand implementation. The paper shows a different equation than the demo [here](https://people.eecs.berkeley.edu/~kjamieson/hyperband.html).
I hope you don't mind if I close the issue.

**Comment by fengzhao860814** [comment]
xx

**Comment by fengzhao860814** [comment]
Hi

whether following problem 2 is resolved,how to solve it?
After the optimizer goes over all configurations it just hangs, waiting for workers:
DEBUG:hpbandster:HBMASTER: Trying to run another job!
DEBUG:hpbandster:job_callback for (19, 0, 0) finished
DEBUG:hpbandster:DISPATCHER: Starting worker discovery
DEBUG:hpbandster:DISPATCHER: Found 8 potential workers, 8 currently in the pool.
DEBUG:hpbandster:DISPATCHER: Finished worker discovery
DEBUG:hpbandster:DISPATCHER: Starting worker discovery
